### ISSUE TITLE
gitignore rules are not prominently documented and ignore checked in files

### ISSUE THREAD

**Issue by ktbarrett** [psf/black#2164]
My `.gitignore` is quite simple. It forces me to explicitly add new files or use `-u`; which is good practice and I don't have to worry about adding a bunch of exclusion rules as things pile up.

```
*
```

However black's "convenient" functionality of using the `.gitignore` to decide which files to exclude *without taking into account whether or not the file is checked in* means I cannot simply apply `black` to my source directory. 

```sh
> black src/
No Python files are present to be formatted. Nothing to do 😴
> rm .gitignore 
> black src/
All done! ✨ 🍰 ✨
6 files left unchanged.
> black --version
black, version 21.4b2
```

This is not the first large project I've seen where "convenient" features eventually weasel their way in. They might help the common case, but can totally defy other user's expectations and are *rarely* documented (\*cough cough\* `setuptools_scm` \*cough\*). It's mad. Resist the temptation please. Code formatting has *nothing* to do with version control. If someone thinks it's important, let it be a plugin and extra: `black[vcs]`

**Comment by ktbarrett** [comment]
I'm not saying it's totally bad feature. My point of view is that introducing a convenience (read optional, but positive) should not be optimizing the common *at the detriment of the uncommon case*. We aren't talking about total program performance here, where taking a small hit somewhere for the advantage of overall performance increase is okay; we are talking about people. 

On second inspection I think the issue is more that the implementation of this feature is incorrect, a file can be excluded while also being version controlled. The proper logic for exclusion is `ignored and not version_controlled`.

Of course I understand this will likely require the project to take on a dependency on `pygit` for a feature that not everyone will use, which isn't ideal. And of course other VCSs go without attention, maybe some user would like SVN support? Should that be added? And they will likely need lots of options to configure those features... With VCS support there is an unbounded interface on the project that will either balloon, or require some arbitrary line in the sand to be drawn that will peeve the users who end up on the opposite side of the line. It's not a good place to be.

If you do not intend to fix the logic, I would do the least intrusive thing, which is option 1. 

I've worked around the issue for now by changing my `.gitignore` to the following. I doubt there will be any Python source files in my project I do not intend to check in if I already haven't.
```
*
# to support black
!*.py
```

> use pre-commit

I am not a fan of modifying source files in pre-commit. I already use `black --check` in pre-commit.

And I hope you understand when I say "you" I don't mean you personally, but you "the project". Thank you for responding.

**Comment by ichard26** [comment]
> convenience (read optional, but positive) should not be optimizing the common at the detriment of the uncommon case. 

I'd argue the opposite to a degree. Now optimizing for the common case and blocking the uncommon cases from ever working? That crosses my line in the sand, but in general, we can't get this perfect. Yes there's the people component to this issue, but it's about tradeoffs here, if there's a bunch of users that are all hitting this issue and find it annoying, sure, I wouldn't mind disrupting most our users and going with No. 1. And that's the next point I want to make, Option 1. is intrusive in its own way since I'm sure plenty of people are now dependant on this feature (including me! - I hope that covers my biases) and yanking it from under them would certainly be annoying and frustrating. Yes it's an one time cost, but I don't think it's worth it.

> The proper logic for exclusion is `ignored and not version_controlled`

I mean, maybe? It totally makes sense to me to have files checked into version control but are still excluded from Black. Maybe the file is Python 2 and can't be formatted by Black. Or it purposefully has bad formatting or even invalid syntax (e.g, our `./tests/data` directory in this repo). 

Finally, I'm going to ping some other maintainers to see what their thoughts are about this /cc @JelleZijlstra @cooperlees

**Comment by ktbarrett** [comment]
One more argument is that `flake8` does not support this feature and just runs forever. Although it doesn't even include reasonable default exclusions like `.tox/`, `.nox/` and `.eggs/`, etc. which might be too far in the other direction.

> I mean, maybe? It totally makes sense to me to have files checked into version control but are still excluded from Black. Maybe the file is Python 2 and can't be formatted by Black.

This is not the purpose of a `.gitignore`. This is what a configuration file is for.

The previous comment was troubling so I decided to try configuring via `pyproject.toml` to set the exclude and include to what I would expect and found that apparently the `.gitignore` rules take precedence over actual configurations! This is really bad...

**Comment by ktbarrett** [comment]
> Because we have --include, I think the most sensible option is to merge gitignore with the value of --exclude; if there's no --exclude, merge it with the default. Values in pyproject.toml basically replace the default value for command line options.

Oof... I absolutely disagree. Precedence is about *overriding* previous configurations. `.gitignore` should be a fallback if exclusions are not set, not a fundamental truth to be merged in. It's essentially turning all setting of exclusions in the configuration file and command line into extra exclusions. However, if you specify `--exclude` on the command line it *overrides* the configuration file, but neither override the `.gitignore`? It's just not clean... Hence why you need the flag you suggested. I think using `.gitignore` while not checking if the file is being version controlled is not great; but this is an even less palatable decision.

**Comment by ktbarrett** [comment]
I think I may have gotten slightly confused on that last bit, I don't think the documentation is clear.

> Exclusions are calculated first, inclusions later.

The first time I reasoned about it as "if inclusion is considered later, that means it has precedence", but what it really means is "files are excluded first and remaining files are then checked for inclusion". Which plainly doesn't make sense if I stopped and thought about what the default patterns were. Still I think the documentation could be made more clear and concise: "exclusion has higher precedence than inclusion."

> What is the scope here? The scope is seemingly growing in size, and I'm losing track of what's being discussed here

Most of what I mentioned was just showcasing how playing around with version control can easily balloon in scope and complexity and it's a rock best left unturned (at least in the base project). But since then I've just been trying to find a way to work around the issue cleanly without having to modify my `.gitignore`. My workflow should not have to change to accommodate a single tool's assumptions.

> What is even the problem?

The next issue is that the `.gitignore` is *always* considered and has precedence over every other rule. I don't think black should be treating `.gitignore` as another configuration file, it was never intended as such. At best it is a convenient default in the case of explicit configuration not existing. If `pyproject.toml` or the commandline receive `exclude` it should *override* `.gitignore`. `extend_exclude` will extend `.gitignore` if no `exclude` is present.

> What solutions are we considering here?

Basically, the comment you linked (https://github.com/psf/black/issues/475#issuecomment-444278828) I think is totally wrong and we should do the opposite. If `exclude` is passed to that function, `gitignore` is not considered.

**Comment by ichard26** [comment]
Thank you very much @ktbarrett! It's pretty late where I am right now so I'll respond later with my proper (counter)arguments of my own (well maybe, you've done a decent job of warming me up to accepting more disruption to clean up the semantics - although please don't take that as "I'm accepting your solution 100%"). I just wanted to say my thanks for clarifying when you didn't have to, I appreciate it a lot. I should do better and try not to get these things wrong as a maintainer, but oh well, I suppose you gotta live and learn sometimes... Even if we two aren't able to reach an acceptable middle ground, you've certainly expanded my view on this topic so thank you for engaging me in some meaningful discussion :)

**Comment by ktbarrett** [comment]
And thank you as well for responding. I've been piecemealing the diagnosis between work, and my comments weren't very well designed. To summarize:

> There is a configuration hierarchy: default/gitignore -> pyproject.toml -> command line.
> `exclude` specified at any level of the hierarchy overrides the previous level.
> `extra-excludes` at any level is additive to the previous level's `exclude`.

Right now gitignore is misbehaving because you can't override it.
